packages/build/tests/blobs_upload/tests.js

```diff
@@ -0,0 +1,89 @@
+import { mkdtemp } from 'node:fs/promises'
+import { tmpdir } from 'node:os'
+import { join } from 'node:path'
+
+import { BlobsServer, getDeployStore } from '@netlify/blobs'
+import { Fixture, normalizeOutput } from '@netlify/testing'
+import test from 'ava'
+import getPort from 'get-port'
+
+const TOKEN = 'test'
+
+test.beforeEach(async (t) => {
+  const port = await getPort()
+  t.context.blobRequestCount = { set: 0, get: 0 }
+
+  t.context.blobServer = new BlobsServer({
+    port,
+    token: TOKEN,
+    directory: await mkdtemp(join(tmpdir(), 'blobs-upload-plugin-')),
+    onRequest: ({ type }) => {
+      t.context.blobRequestCount[type] = (t.context.blobRequestCount[type] || 0) + 1
+    },
+  })
+
+  await t.context.blobServer.start()
+
+  process.env.NETLIFY_BLOBS_CONTEXT = Buffer.from(
+    JSON.stringify({
+      edgeURL: `http://localhost:${port}`,
+      token: TOKEN, // This is for tests, during build this is passed as argument
+    }),
+  ).toString('base64')
+})
+
+test.afterEach.always(async (t) => {
+  await t.context.blobServer.stop()
+  delete process.env.NETLIFY_BLOBS_CONTEXT
+})
+
+test("blobs upload, don't run when deploy id is provided and no files in directory", async (t) => {
+  const output = await new Fixture('./fixtures/src_empty').withFlags({ deployId: 'abc123' }).runWithBuild()
+
+  t.is(t.context.blobRequestCount.set, 0)
+
+  t.snapshot(normalizeOutput(output))
+})
+
+test("blobs upload, don't run when there are files but deploy id is not provided", async (t) => {
+  const output = await new Fixture('./fixtures/src_with_blobs').withFlags({}).runWithBuild()
+
+  t.is(t.context.blobRequestCount.set, 0)
+
+  t.snapshot(normalizeOutput(output))
+})
+
+test.serial('blobs upload, uploads files to deploy store', async (t) => {
+  const output = await new Fixture('./fixtures/src_with_blobs')
+    .withFlags({ deployId: 'abc123', siteId: 'test' })
+    .runWithBuild()
+
+  t.is(t.context.blobRequestCount.set, 3)
+
+  const store = getDeployStore({ deployID: 'abc123', siteID: 'test' })
+
+  let blob
+  blob = await store.getWithMetadata('something.txt')
+  t.is(blob.data, 'some value')
+  t.deepEqual(blob.metadata, {})
+
+  blob = await store.getWithMetadata('with-metadata.txt')
+  t.is(blob.data, 'another value')
+  t.deepEqual(blob.metadata, { meta: 'data', number: 1234 })
+
+  blob = await store.getWithMetadata('nested/file.txt')
+  t.is(blob.data, 'file value')
+  t.deepEqual(blob.metadata, { some: 'metadata' })
```

Comment:
nit: I would avoid reassigning if we can avoid it. This boils down to preference, and is especially inconsequential in a test, so by all means feel free to take or leave.

```suggestion
const blob1 = await store.getWithMetadata('something.txt')
t.is(blob1.data, 'some value')
t.deepEqual(blob1.metadata, {})

const blob2 = await store.getWithMetadata('with-metadata.txt')
t.is(blob2.data, 'another value')
t.deepEqual(blob2.metadata, { meta: 'data', number: 1234 })

const blob3 = await store.getWithMetadata('nested/file.txt')
t.is(blob3.data, 'file value')
t.deepEqual(blob3.metadata, { some: 'metadata' })
```